polygon-digital-carbon/src/RetirementHandler.ts

```diff
@@ -195,3 +201,98 @@ export function saveICRRetirement(event: RetiredVintage): void {
+
+  let credit = loadOrCreateCarbonCredit(event.params.fromToken, 'C3', null)
+  credit.save()
+
+  // ensure accounts are created for all addresses
+  loadOrCreateAccount(event.params.beneficiary)
+  let sender = loadOrCreateAccount(event.transaction.from)
+  let senderAddress = event.transaction.from
+
+  let retireId = senderAddress.concatI32(sender.totalRetirements)
+
+  saveRetire(
+    retireId,
+    credit.id,
+    ZERO_ADDRESS,
+    'OTHER',
+    event.params.amount,
+    event.params.beneficiary,
+    '',
+    senderAddress,
+    '',
+    event.block.timestamp,
+    event.transaction.hash,
+    'C3'
+  )
+
+  let retire = loadRetire(retireId)
+
+  let requestId = getC3OffsetRequestId(event.params.fromToken, event.params.index)
+  let request = loadOrCreateC3OffsetBridgeRequest(requestId)
+
+  request.status = 'REQUESTED'
+  request.index = event.params.index
+  request.retire = retireId
+  request.provenance = retire.provenance
+  request.save()
+
+  retire.c3OffsetRequest = requestId
+  retire.save()
+
+  incrementAccountRetirements(senderAddress)
+}
+
+export function completeC3OffsetRequest(event: EndAsyncToken): void {
+  log.info('completeC3OffsetRequest event fired {}', [event.transaction.hash.toHexString()])
+
+  loadOrCreateAccount(event.transaction.from)
+
+  let requestId = getC3OffsetRequestId(event.params.fromToken, event.params.index)
+  let request = loadC3OffsetBridgeRequest(requestId)
+
+  if (request == null) {
+    log.error('No C3OffsetRequest found for retireId: {} hash: {}', [
+      // retireId.toHexString(),
+      event.transaction.hash.toHexString(),
+    ])
+    return
+  } else {
+    if (request.status == 'REQUESTED') {
+      let c3OffsetNftContract = C3OffsetNFT.bind(C3_VERIFIED_CARBON_UNITS_OFFSET)
+
+      let tokenURICall = c3OffsetNftContract.try_tokenURI(event.params.nftIndex)
+
+      request.c3OffsetNftIndex = event.params.nftIndex
+      if (tokenURICall.reverted) {
+        log.error('tokenURI call reverted for NFT index {}', [event.params.nftIndex.toString()])
+      } else {
+        let tokenURI = tokenURICall.value
+        if (tokenURI == null || tokenURI == '') {
+          let safeguard = TokenURISafeguard.load('safeguard')
+          if (safeguard == null) {
+            safeguard = new TokenURISafeguard('safeguard')
+            safeguard.requestsWithoutURI = []
+            safeguard.save()
+          }
+          let requestsArray = safeguard.requestsWithoutURI
+          requestsArray.push(requestId)
+          safeguard.requestsWithoutURI = requestsArray
+          safeguard.save()
+          log.error('Retrieved tokenURI is null or empty for nft index {}', [event.params.nftIndex.toString()])
+        }
+        request.tokenURI = tokenURI
+      }
+
+      request.status = 'FINALIZED'
+      request.save()
+      /** decrement account retirements because the retire is double counted in VCUOMinted */
+      // decrementAccountRetirements(Address.fromBytes(sender.id))
```

Comment:
Does this need to be implemented before merging? Seems important to avoid double counting...

Reply:
Ended up not need this with the current implementation, removed.